### DISCUSSION TITLE
custom initrd to mount an overlayfs

### DISCUSSION THREAD

**Discussion by k-e-r-n-e-l-p-a-n-i-c** [firecracker-microvm/firecracker#3345] [Q&A]
Hi All,

I intend to use a writable squashfs  filesystem as rootfs for my microvm.

To do so I have a custom initrd archive with an init script that does the following:

1. Mount /dev/vda on a temporary mount point
2. Create /overlay, /lower, /upper and /work directories for overlayfs
3. Move the already mounted /dev/vda to /lower directory of overlayfs
4. Mount overlayfs at /overlay
5. cd to /overlay and switch_root to '/'. Execute a custom /sbin/init script.

The above script works fine and I am able to mount the squashfs with a writable layer over it.  However, I would like to understand if this is a recommended process or am I doing it incorrectly?   Please share your opinions/guidance.

Regards,
Arun Prasad

**Comment by cipriancraciun** [comment]
I don't think the `firecracker` project is the best place to be asking this question, because the focus of the project is mainly with the virtualization and less with what happens inside the VM after it boots.  (Although I'm not a developer of this project, thus I might be wrong.)

Instead I would suggest looking at various Linux distributions (especially ArchLinux) on how they handle this use-case.  Because the early Linux init, and especially the handling of `rootfs` might be a bit tricky, I would pay especially attention to the `initramfs` of many of these distributions (the `switch_root` magic happens in there), or how alternative init systems (like `runit`, `s6` or `finit`) handle this.

**Reply by k-e-r-n-e-l-p-a-n-i-c** [reply]
@cipriancraciun  thanks. Found some valuable information at  [ArchLinux ](https://gitlab.archlinux.org/archlinux/mkinitcpio/mkinitcpio) repository. Will dig deeper to get more clarity.

**Comment by zulinx86** [comment]
As @cipriancraciun mentioned, it does not look a question for firecracker but linux administration. But we have a script to generate a rootfs we're using in tests [here](https://github.com/firecracker-microvm/firecracker/blob/main/resources/rebuild.sh).

@k-e-r-n-e-l-p-a-n-i-c seems to have already found valuable information, closing this discussion.
